### DISCUSSION TITLE
how do I invert a search with multiple patterns?

### DISCUSSION THREAD

**Discussion by zz5678** [BurntSushi/ripgrep#2628] [Q&A] [answered]
For the following to be processed:
```
abc
efg
123
```
my command is:
```
rg -v ab -v 12
```

This command doesn't actually work, but the ChatGPT answer works like this.

Is there a feature like this? But I'm using it in wrong way?

![Screenshot_20231013142128](https://github.com/BurntSushi/ripgrep/assets/67672527/778597c0-6971-41b7-8927-90fa12123808)

**Comment by s-p-turner** [comment, marked as answer]
Fwiw, two alternative ways to achieve what you want would be either:

rg -v ab "input-file" | rg -v 12

or (more efficient):

rg -v 'ab|12' "input-file"

**Comment by zz5678** [comment]
> Fwiw, two alternative ways to achieve what you want would be either:
> 
> rg -v ab "input-file" | rg -v 12
> 
> or (more efficient):
> 
> rg -v 'ab|12' "input-file"

Yes, the second method is more efficient, but it doesn't seem to appear in the man page of ripgrep

One question is, who told ChatGPT about the Way that doesn't exist????

**Comment by BurntSushi** [comment]
(I've converted this to a Discussion since this is more Q&A than anything else. I've also marked @s-p-turner's response as the answer. I'll provide a bit more context here.)

ripgrep's man page doesn't contain all possible forms of composition. Instead, it documents how to compose things. For example, multiple `-e/--regexp` flags can be given. The `-v/--invert-match` flag is a switch that applies to the overall search. The man page documents it as a flag that does _not_ take a value, so your use here is erroneous.

I myself will use both `rg -v 'foo|bar'` or `rg -v foo | rg -v bar`, depending on the occasion. It just depends on what I'm doing. It's true that the latter is technically likely less efficient than the former, but it's unlikely to be noticeable in the vast majority of cases. Especially since pipelines like this tend to reduce the corpus size at each step, and thus, the marginal cost of an additional command in the pipeline gets lower and lower as the search is refined. Just write what is natural to you. And if it's too slow, then you can start looking at ways to speed it up.

As for ChatGPT, I don't know why it says the things it does. If you want to talk ChatGPT, please do that somewhere else.
